### ISSUE TITLE
DistributedDataParallel works better on NVlink

### ISSUE THREAD

**Issue by yaroslavvb** [kimiyoung/transformer-xl#66]
Replacing DataParallel -> DistributedDataParallel makes step time go from from 350ms to 260ms on p3.8xlarge machine for the `run_wt103_base.sh` example ([example cl](https://github.com/cybertronai/transformer-xl/commit/1a35238763b6cd5e73c488d9795e8a338418b67b))

**Comment by zihangdai** [comment]
Thanks a lot for the cl. We will test it a little bit before merging it into the repo.

**Comment by yaroslavvb** [comment]
The CL doesn't really work standalone and is there just as an example. The important parts are:
- separate script to call `python -m torch.distributed.launch`
- torch.cuda.set_device(args.local_rank) before any `.to(device)` calls
- `dist.init_process_group(backend='nccl',`

**Comment by zihangdai** [comment]
Thank you for the instructions. I happened to be familiar with DistributedDataParallel but not the AWS part. I'll look into it when time allows. Thanks again.

**Comment by yaroslavvb** [comment]
One thing I noticed is that the network uses LayerNorm, with DistributedDataParallel, the effective batch size is 15 rather than 60, so LayerNorm normalizes a bit differently. It seems the batch-60 layer norm is slightly better. Orange is original code, green is DistributedDataParallel version

<img width="363" alt="Screenshot 2019-04-22 19 19 06" src="https://user-images.githubusercontent.com/23068/56547954-a788ab00-6533-11e9-8fc1-c66b70ca3924.png">

Also, since all GPUs run an identical training loop, it needs some logic to avoid having each process read identical data, something like this

```
    batch = -1
    train_iter = iter(train_iter)
    while True:
        # hack to scatter data across processes
        batch += 1
        inputs = []
        try:
            for i in range(max_rank):
                inputs.append(next(train_iter))
        except StopIteration:
            break
        
        data, target, seq_len = inputs[global_rank]
```

**Comment by yaroslavvb** [comment]
The throughput is 30% better for `run_wt103_base.sh`, but I'm still trying to figure out why sample efficiency is about 30% worse. Because DDP creates a separate copy of model for each GPU, things like Batch Normalization will essentially work with batch size=15 rather than batch size 60. However, I don't see anything in the model tied to batch size.

Does the order of data feeding matter? Right now I'm doing something [very simple](https://github.com/cybertronai/transformer-xl/blob/a8675689dd95026475249312ccdc477079f0b929/pytorch/train.py#L634), GPU-0 gets examples 0, 4, 8. GPU-1 gets examples 1, 5, 9 etc, would this be a problem?



![Screenshot 2019-04-23 09 46 18](https://user-images.githubusercontent.com/23068/56600144-e6a71280-65ac-11e9-81a6-ac25dbe84605.png)

**Comment by zihangdai** [comment]
That's weird. As far as I can think of, there is nothing tied to the batch size. As for data feeding, the only important thing is to keep the temporal order correct. Another thing is to make sure each device handles its own memory.

**Comment by yaroslavvb** [comment]
Ah, it must be the temporal correctness issue -- setting memory to 0 in run_wt103_base.sh also gives about 30% penalty in sample efficiency, so "naive DDP" has almost same curve as "DP with no memory"

<img width="638" alt="Screenshot 2019-04-23 10 30 15" src="https://user-images.githubusercontent.com/23068/56602651-d42fd780-65b2-11e9-98e8-b160e2d2c4d1.png">

**Comment by 8enmann** [comment]
Fixed https://github.com/cybertronai/transformer-xl/commit/8120da645efccaa489c45c60382d25c8d16b8ef3
![2019-04-23](https://user-images.githubusercontent.com/1021104/56698920-75e52080-66a8-11e9-84da-a711fde00116.png)
blue line is original, lower red line is with sequences preserved

**Comment by yaroslavvb** [comment]
BTW, the end-to-end solution is [here](https://github.com/cybertronai/transformer-xl). It's about 2.6x faster by also adding proper fp16 mode
